filetracker/servers/storage.py

```diff
@@ -108,62 +109,53 @@ def store(self, name, data, version, size=0, compressed=False, digest=None):
             if _path_exists(link_path) and _file_version(link_path) > version:
                 return _file_version(link_path)
 
-            # Path to temporary file that may be created in some cases.
-            temp_file_path = None
-
-            if digest is None:
-                # Write data to temp file and calculate hash.
-                temp_file_fd, temp_file_path = tempfile.mkstemp()
-                temp_file = os.fdopen(temp_file_fd, 'wb')
-                _copy_stream(data, temp_file, size)
-                temp_file.close()
-
-                if compressed:
-                    # If data was already compressed, we have to decompress it
-                    # before calculating the digest.
-                    with gzip.open(temp_file_path, 'rb') as compressed_file:
-                        digest = file_digest(compressed_file)
-                else:
-                    digest = file_digest(temp_file_path)
+            # data is managed by contents now, and shouldn't be used directly
+            with _InputStreamWrapper(data, size) as contents:
+                if digest is None or logical_size is None:
+                    contents.save()
+                    if compressed:
+                        # This shouldn't occur if the request came from a proper
+                        # filetracker client, so we don't care if it's slow.
+                        with gzip.open(
+                                contents.current_path, 'rb') as decompressed:
+                            digest = file_digest(decompressed)
+                        with gzip.open(
+                                contents.current_path, 'rb') as decompressed:
+                            logical_size = _read_stream_for_size(decompressed)
+                    else:
+                        digest = file_digest(contents.current_path)
+                        logical_size = os.stat(contents.current_path).st_size
+
+                blob_path = self._blob_path(digest)
+                
+                with self._lock_blob_with_txn(digest) as txn:
+                    digest_bytes = digest.encode()
 
-            blob_path = self._blob_path(digest)
-            
-            with self._lock_blob_with_txn(digest) as txn:
-                digest_bytes = digest.encode('utf8')
-                try:
                     link_count = int(self.db.get(digest_bytes, 0, txn=txn))
-                except KeyError:
-                    link_count = 0
+                    new_count = str(link_count + 1).encode()
+                    self.db.put(digest_bytes, new_count, txn=txn)
 
-                new_count = str(link_count + 1).encode('utf8')
-                self.db.put(digest_bytes, new_count, txn=txn)
+                    # Create a new blob if this isn't a duplicate.
+                    if link_count == 0:
+                        _create_file_dirs(blob_path)
+                        self.db.put('{}:logical_size'.format(digest).encode(),
```

Comment:
`digest:logical_size` is added to the DB here and it's never removed.
I think it should be removed [here](https://github.com/sio2project/filetracker/pull/41/files#diff-588c7e73493760c7e36630cdbffa27afR207) when the last link to the blob has been removed and the reference counter is being removed.